### ISSUE TITLE
Usage of errorneous constant can be omitted on nightly and beta

### ISSUE THREAD

**Issue by AnthonyMikh** [rust-lang/rust#67083]
Consider the following code:
```rust
trait ZeroSized: Sized {
    #[deny(const_err)]
    const I_AM_ZERO_SIZED: ();
    fn requires_zero_size(self);
}

impl<T: Sized> ZeroSized for T {
    const I_AM_ZERO_SIZED: ()  = [()][std::mem::size_of::<Self>()];
    fn requires_zero_size(self) {
        #![deny(const_err)]
        let () = Self::I_AM_ZERO_SIZED;
        println!("requires_zero_size called");
    }
}

fn main() {
    ().requires_zero_size();
    42_u32.requires_zero_size();
}
```
The intent of code is to prevent `requires_zero_size` from being called on types with non-zero size (duh). This goal is achieved on stable: trying to compile this code with [stable compiler (1.39.0)](https://play.rust-lang.org/?version=stable&gist=75df6069bcc5b24744d48ae57c552d34) gives the following error:
```
error: any use of this value will cause an error
 --> src/main.rs:8:34
  |
8 |     const I_AM_ZERO_SIZED: ()  = [()][std::mem::size_of::<Self>()];
  |     -----------------------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-
  |                                  |
  |                                  index out of bounds: the len is 1 but the index is 4
  |
  = note: `#[deny(const_err)]` on by default
```
I expected the same with some newer unstable versiosns. However, this code compiles just fine on [latest beta (2019-12-02 1463fedbddfc0ee087ec)](https://play.rust-lang.org/?version=beta&gist=75df6069bcc5b24744d48ae57c552d34) and [latest nightly (2019-12-05 710a362dc7634fce4288)](https://play.rust-lang.org/?version=nightly&gist=75df6069bcc5b24744d48ae57c552d34) at the moment.

Compiling in debug vs release gives no difference in behaviour.

**Comment by Centril** [comment]
cc @wesleywiser @oli-obk

**Comment by oli-obk** [comment]
Oh oh. Investigating...

**Comment by oli-obk** [comment]
Wild guess: const prop based on miri does not actually read from zst constants... and confirmed by making the constant's type i32 instead of ()

**Comment by AnthonyMikh** [comment]
Interesting: just changing type to `i32` is not sufficient, you should actually bind the value to name. This one fails to compile:
```rust
trait ZeroSized: Sized {
    #[deny(const_err)]
    const I_AM_ZERO_SIZED: i32;
    fn requires_zero_size(self);
}

impl<T: Sized> ZeroSized for T {
    const I_AM_ZERO_SIZED: i32  = [0][std::mem::size_of::<Self>()];
    fn requires_zero_size(self) {
        #![deny(const_err)]
        let _unused = Self::I_AM_ZERO_SIZED;
        println!("requires_zero_size called");
    }
}

fn main() {
    ().requires_zero_size();
    42_u32.requires_zero_size();
}
```
Making a slight change:
```
11 -        let _unused = Self::I_AM_ZERO_SIZED;
11 +         let _ = Self::I_AM_ZERO_SIZED;
```
...And this compiles again on beta and nightly.

**Comment by oli-obk** [comment]
It's not const prop. Const prop doesn't know about this yet. If you allow all the lints on stable, you get a segfault in the running program. I think monomorphization needs to hard error on invalid constants. Trying out now

**Comment by RalfJung** [comment]
> The intent of code is to prevent requires_zero_size from being called on types with non-zero size (duh). This goal is achieved on stable

Notice that seems quite fragile though, as this is only prevented by a deny-by-default *lint*. With `--cap-lints` (as it is used for dependencies) or when someone uses your crate with `#[allow(const_err)]`, this code will compile on stable already.

**Comment by oli-obk** [comment]
Usage of a broken constant is not just a lint, it is a hard error in monomorphic code. What you are thinking about is promotion, where we never report a hard error. Using a constant whose evaluation would fail is a hard error, it's just that declaring said constant and not using it will just cause warnings.

**Comment by pnkfelix** [comment]
triage: P-high, removing nomination, assigning to @oli-obk

**Comment by pnkfelix** [comment]
fixed by beta backport  PR #67297 which was backport of PR #67134
